swift-sdk/IterableActionRunner.swift

```diff
@@ -33,58 +29,120 @@ class AppUrlOpener : UrlOpenerProtocol {
     }
-    
-    private func open(url: URL, inContext context: IterableActionContext) -> Bool {
-        if urlDelegate?.handle(iterableURL:url, inContext: context) == true {
-            return true
-        }
 
-        guard context.source == .push else {
-            // only open urls for push, leave others to delegate
+    // return true if the action is handled by the calling app either by opening a url or performing a custom action.
+    private static func callExternalHandlers(action: IterableAction,
+                        from source: IterableActionSource,
+                        urlHandler: UrlHandler? = nil,
+                        customActionHandler: CustomActionHandler? = nil) -> Bool {
+        let actionType = detectActionType(fromAction: action)
+        switch(actionType) {
+        case .noop:
             return false
+        case .openUrl(let url):
+            if urlHandler?(url) == true {
+                return true
+            } else {
+                return false
+            }
+        case .customAction(let type):
+            if customActionHandler?(type) == true {
+                return true
+            } else {
+                return false
+            }
         }
-        
-        if let scheme = url.scheme, scheme == "http" || scheme == "https" {
-            urlOpener.open(url: url)
+    }
+
+    // MARK: Private
+    private static func shouldOpenUrl(url: URL, from source: IterableActionSource) -> Bool {
+        if source == .push, let scheme = url.scheme, (scheme == "http" || scheme == "https") {
             return true
         } else {
             return false
         }
     }
+
+    private enum ActionType {
+        case openUrl(URL)
+        case customAction(String)
+        case noop
+    }
     
-    private func callCustomActionIfSpecified(action: IterableAction, inContext context: IterableActionContext) -> Bool {
-        guard IterableUtil.isNotNullOrEmpty(string: action.type) else {
-            return false
+    // What type of action needs to be performed.
+    private static func detectActionType(fromAction action: IterableAction) -> ActionType {
+        if action.isOpenUrl() {
+            if let urlString = action.data, let url = URL(string: urlString) {
+                return .openUrl(url)
+            } else {
+                ITBError("Could not create url from action: \(action)")
+                return .noop
+            }
+        } else {
+            if IterableUtil.isNullOrEmpty(string: action.type) {
+                return .noop
+            } else {
+                return .customAction(action.type)
+            }
         }
-        guard let customActionDelegate = customActionDelegate else {
-            return false
+    }
+
+    // MARK: Helper Utility Functions
+    // converts from IterableURLDelegate to UrlHandler
+    static func urlHandler(fromUrlDelegate urlDelegate: IterableURLDelegate?, inContext context: IterableActionContext) -> UrlHandler {
```

Comment:
I don't have a strong opinion on this, but do `UrlHandler`s offer any isolation vs. `IterableURLDelegate`s? It looks like `IterableActionContext` and `IterableURLDelegate` are still used in this file.

Reply:
I think they do. These static methods can be moved to any other appropriate place later.

Reply:
I will move them to IterableUtil.